zerver/migrations/0699_fix_user_role_system_groups.py

```diff
@@ -0,0 +1,191 @@
+        }
+        for value_dict in (
+            UserProfile.objects.filter(realm=realm, is_bot=False, is_active=True)
+            .values("role")
+            .annotate(Count("role"))
+        ):
+            human_counts[str(value_dict["role"])] = value_dict["role__count"]
+        bot_count = UserProfile.objects.filter(realm=realm, is_bot=True, is_active=True).count()
+        return {
+            ROLE_COUNT_HUMANS: human_counts,
+            ROLE_COUNT_BOTS: bot_count,
+        }
+
+    def fix_user_memberships_if_needed(user_profile: Any) -> None:
+        realm = user_profile.realm
+
+        guest_group = get_system_user_group_by_name(SystemGroups.EVERYONE, realm.id)
+        member_group = get_system_user_group_by_name(SystemGroups.MEMBERS, realm.id)
+        moderator_group = get_system_user_group_by_name(SystemGroups.MODERATORS, realm.id)
+        admin_group = get_system_user_group_by_name(SystemGroups.ADMINISTRATORS, realm.id)
+        owner_group = get_system_user_group_by_name(SystemGroups.OWNERS, realm.id)
+
+        group_id_to_group_name = {
+            guest_group.id: "Guests",
+            member_group.id: "Members",
+            moderator_group.id: "Moderators",
+            admin_group.id: "Administrators",
+            owner_group.id: "Owners",
+        }
+        role_to_group = {
+            ROLE_GUEST: guest_group,
+            ROLE_MEMBER: member_group,
+            ROLE_MODERATOR: moderator_group,
+            ROLE_REALM_ADMINISTRATOR: admin_group,
+            ROLE_REALM_OWNER: owner_group,
+        }
+        group_id_to_role = {v.id: k for k, v in role_to_group.items()}
+
+        role_group_ids = [
+            guest_group.id,
+            member_group.id,
+            moderator_group.id,
+            admin_group.id,
+            owner_group.id,
+        ]
+
+        role = user_profile.role
+        group_implied_by_role = role_to_group[role]
+
+        direct_memberships = UserGroupMembership.objects.filter(
+            user_profile=user_profile, user_group_id__in=role_group_ids
+        )
+        if len(direct_memberships) == 0:
+            print(
+                f"User {user_profile.id} has no role group memberships. This is unexpected. Skipping."
+            )
+            return
+
+        if len(direct_memberships) > 1:
+            group_names = [group_id_to_group_name[m.user_group_id] for m in direct_memberships]
+            print(
+                f"User {user_profile.id} has more than one role group membership: {group_names}. "
+                f"Expected group based on role value of {role}: {group_id_to_group_name[group_implied_by_role.id]}. Skipping."
+            )
+            return
+
+        assert len(direct_memberships) == 1
+        role_membership = direct_memberships[0]
+
+        if role_membership.user_group_id == group_implied_by_role.id:
+            # This user's state is correct.
+            return
+
+        print(f"Fixing user {user_profile.id}...")
```

Comment:
Is the intent for this one to not include a newline so that the `ok` isn't on a line by itself?

Reply:
It wasn't, but yeah that's much cleaner output. fixed with
```
        print(f"Fixing user {user_profile.id}... ", end="")
```

Reply:
Nevermind, with the new bulk-queries approaching this is now gone, as we no longer fix users one by one.
Just doing
```
        print(f"User {user_profile.id} will be fixed.")
```

now.